Mathlib/Control/Monad/Free.lean

```diff
@@ -0,0 +1,577 @@
+  Haskell Symposium 2015.
+* [MilewskiDao] Bartosz Milewski. *The Dao of Functional Programming*. 2025.
+
+## Tags
+
+Free monad, state monad, writer monad, continuation monad
+-/
+
+/-- The Free monad over a type constructor `f`.
+
+A `FreeM f a` is a tree of operations from the type constructor `f`, with leaves of type `a`.
+It has two constructors: `pure` for wrapping a value of type `a`, and `liftBind` for
+representing an operation from `f` followed by a continuation.
+
+This construction provides a free monad for any type constructor `f`, allowing for composable
+effect descriptions that can be interpreted later. Unlike the traditional free monad,
+this does not require `f` to be a functor. -/
+inductive FreeM.{u, v, w} (f : Type u → Type v) (α : Type w) where
+  | protected pure : α → FreeM f α
+  | liftBind {ι : Type u} (op : f ι) (cont : ι → FreeM f α) : FreeM f α
+
+/-
+Disable simpNF lints for auto-generated constructor lemmas, as they don't follow simp normal
+form patterns. The LHS of these lemmas use `FreeM.pure` which simplifies to `pure` via
+`pure_eq_pure`.
+-/
+attribute [nolint simpNF] FreeM.pure.sizeOf_spec FreeM.pure.injEq
+
+universe u v w
+
+namespace FreeM
+
+/-- Map a function over a `FreeM` monad. -/
+def map {α β : Type w} {F : Type u → Type v} (f : α → β) : FreeM F α → FreeM F β
+  | .pure a => .pure (f a)
+  | .liftBind op cont => .liftBind op (fun z => FreeM.map f (cont z))
+
+instance {F : Type u → Type v} : Functor (FreeM F) where
+  map := .map
+
+instance {F : Type u → Type v} : LawfulFunctor (FreeM F) where
+  map_const := rfl
+  id_map x := by
+    simp [Functor.map]
+    induction x with
+    | pure a => rfl
+    | liftBind op cont ih =>
+      simp [map, ih]
+  comp_map g h x := by
+    simp [Functor.map]
+    induction x with
+    | pure a => rfl
+    | liftBind op cont ih => simp [FreeM.map, ih]
+
+/-- Bind operation for the `FreeM` monad. -/
+protected def bind {a b : Type w} {F : Type u → Type v} (x : FreeM F a) (f : a → FreeM F b) :
+  FreeM F b :=
+  match x with
+  | .pure a => f a
+  | .liftBind op cont => .liftBind op (fun z => FreeM.bind  (cont z) f)
+
+/-- Lift an operation from the effect signature `f` into the `FreeM f` monad. -/
+@[simp]
+def lift {F : Type u → Type v} {ι : Type u} (op : F ι) : FreeM F ι :=
+  .liftBind op FreeM.pure
+
+instance {F : Type u → Type v} : Monad (FreeM F) where
+  pure := .pure
+  bind := .bind
+
+@[simp]
+lemma pure_eq_pure {F : Type u → Type v} {α : Type w} (a : α) :
+    (.pure a : FreeM F α) = pure a := rfl
+
+/-- The `liftBind` constructor is semantically equivalent to `(lift op) >>= cont`. -/
+lemma liftBind_eq_lift_bind {F : Type u → Type v} {ι : Type u} {α : Type u}
+    (op : F ι) (cont : ι → FreeM F α) :
+    .liftBind op cont = (lift op) >>= cont := by rfl
+
+instance {F : Type u → Type v} : LawfulMonad (FreeM F) := LawfulMonad.mk'
+  (bind_pure_comp := by {
+    intro _ _ f x;
```

Comment:
`{` is pretty unidiomatic; better as
```suggestion
  (bind_pure_comp := fun _ _ f x => by
```
what made you use the `{}` syntax in the first place?

Reply:
See also https://github.com/leanprover-community/mathlib4/pull/25491#discussion_r2146802389